aptos-move/block-executor/src/txn_provider/blocking_txns_provider.rs

```diff
@@ -0,0 +1,75 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::txn_provider::TxnProvider;
+use aptos_mvhashmap::types::TxnIndex;
+use aptos_types::transaction::BlockExecutableTransaction as Transaction;
+use std::sync::{Arc, Condvar, Mutex};
+
+#[allow(dead_code)]
+pub enum BlockingTransactionStatus<T: Transaction> {
+    Ready(Arc<T>),
+    Waiting,
+}
+
+pub struct BlockingTransaction<T: Transaction> {
+    pub txn: Mutex<BlockingTransactionStatus<T>>,
+    pub cvar: Condvar,
+}
+
+#[allow(dead_code)]
+impl<T: Transaction> BlockingTransaction<T> {
+    pub fn new() -> Self {
+        Self {
+            txn: Mutex::new(BlockingTransactionStatus::Waiting),
+            cvar: Condvar::new(),
+        }
+    }
+}
+
+pub struct BlockingTxnsProvider<T: Transaction> {
+    txns: Vec<BlockingTransaction<T>>,
+}
+
+#[allow(dead_code)]
+impl<T: Transaction> BlockingTxnsProvider<T> {
+    pub fn new(txns: Vec<BlockingTransaction<T>>) -> Self {
+        Self { txns }
+    }
+
+    pub fn set_txn(&self, idx: TxnIndex, txn: T) {
+        let blocking_txn = &self.txns[idx as usize];
+        let (lock, cvar) = (&blocking_txn.txn, &blocking_txn.cvar);
+        let mut status = lock.lock().unwrap();
+        match &*status {
+            BlockingTransactionStatus::Waiting => {
+                *status = BlockingTransactionStatus::Ready(Arc::new(txn));
+                cvar.notify_all();
+            },
+            BlockingTransactionStatus::Ready(_) => {
+                panic!("Trying to add a txn that is already present");
+            },
+        }
+    }
+}
+
+impl<T: Transaction> TxnProvider<T> for BlockingTxnsProvider<T> {
+    fn num_txns(&self) -> usize {
+        self.txns.len()
+    }
+
+    fn get_txn(&self, idx: TxnIndex) -> Arc<T> {
+        let txn = &self.txns[idx as usize];
+        let mut status = txn.txn.lock().unwrap();
+        match *status {
+            BlockingTransactionStatus::Ready(ref txn) => txn.clone(),
+            BlockingTransactionStatus::Waiting => {
+                status = txn.cvar.wait(status).unwrap();
```

Comment:
from the documentation of condvar wait: 
> Note that this function is susceptible to spurious wakeups. Condition variables normally have a boolean predicate associated with them, and the predicate must always be checked each time this function returns to protect against spurious wakeups.
The typical pattern is to do it in a loop, we have an example here:
https://github.com/aptos-labs/aptos-core/blob/5ad601cc4bb5adeb086721a4d57ca3014119c77f/aptos-move/block-executor/src/view.rs#L427
(DependencyCondvar is defined here: https://github.com/aptos-labs/aptos-core/blob/5ad601cc4bb5adeb086721a4d57ca3014119c77f/aptos-move/block-executor/src/scheduler.rs#L63)

So I think what we can do here is have separate condvar with a status lock, use in a loop, and separately an ExplicitSyncWrapper for the transaction itself, then it might also allow us to have an interface for naked &.
An Arc isn't really a big issue but in general I want us to try to reduce reliance on them in cases they aren't actually needed (lifetime guaranteed and reference counting not needed), and this seems like a good example of such a case.

Reply:
@ziaptos: minor question here out of curiosity, if it's not too easy to change to non-arc, would it be better to return &Arc?  (or triopmhe arc?)